match_tpl.h

```diff
@@ -89,50 +117,80 @@ int32_t LONGEST_MATCH(deflate_state *const s, Pos cur_match) {
          * is limited to the lookahead, so the output of deflate is not
          * affected by the uninitialized values.
          */
-        cont = 1;
-        do {
-            match = window + cur_match;
-            if (LIKELY(*(bestcmp_t *)(match+best_len-1) != scan_end ||
-                       *(bestcmp_t *)(match) != scan_start)) {
-                if ((cur_match = prev[cur_match & wmask]) > limit && --chain_length != 0) {
-                    continue;
-                }
-                cont = 0;
+#ifdef UNALIGNED_OK
+        if (best_len < sizeof(uint32_t)) {
+            for (;;) {
+                match = window + cur_match;
+                if (UNLIKELY(*(uint16_t *)(match+offset) == (uint16_t)scan_end &&
+                             *(uint16_t *)(match) == (uint16_t)scan_start))
+                    break;
+                GOTO_NEXT_CHAIN;
             }
-            break;
-        } while (1);
-
-        if (!cont)
-            break;
-
-#if MIN_MATCH >= 2 && defined(UNALIGNED_OK)
-        len = COMPARE256(scan+2, match+2) + 2;
+        }
+#  ifdef UNALIGNED64_OK
+        else if (best_len >= sizeof(uint64_t)) {
+            for (;;) {
+                match = window + cur_match;
+                if (UNLIKELY(*(uint64_t *)(match+offset) == (uint64_t)scan_end &&
+                             *(uint64_t *)(match) == (uint64_t)scan_start))
+                    break;
+                GOTO_NEXT_CHAIN;
+            }
+        }
+#  endif
+        else {
+            for (;;) {
+                match = window + cur_match;
+                if (UNLIKELY(*(uint32_t *)(match+offset) == (uint32_t)scan_end &&
+                             *(uint32_t *)(match) == (uint32_t)scan_start))
+                    break;
+                GOTO_NEXT_CHAIN;
+            }
+        }
 #else
-        len = COMPARE258(scan, match);
+        for (;;) {
+            match = window + cur_match;
+            if (UNLIKELY(match[offset] == scan_end && match[offset+1] == scan_end0 &&
+                         match[0] == scan[0] && match[1] == scan[1]))
+                break;
+            GOTO_NEXT_CHAIN;
+        }
 #endif
-
+        len = COMPARE256(scan+2, match+2) + 2;
         Assert(scan+len <= window+(unsigned)(s->window_size-1), "wild scan");
 
         if (len > best_len) {
             s->match_start = cur_match;
             best_len = len;
-            if (len >= nice_match)
+            if (best_len >= nice_match)
                 break;
-            scan_end = *(bestcmp_t *)(scan+best_len-1);
+            offset = best_len-1;
+#ifdef UNALIGNED_OK
+            if (best_len >= sizeof(uint32_t)) {
```

Comment:
..while here we check for larger than sizeof(unit32_t).

I have not read this code to the point of understanding yet, but this stood out a bit, two questions.
- Should these have been the same?
- If not, why is this case before the UNALIGNED64_OK case, and why is it doing both and not just one of them?
I am probably just missing something here, since it looks deliberate, but it is currently confusing me since there also are not any comments to detail the tricks. I just don't quite see why offset should be best_len-7.

Reply:
bestlen-7 actually means bytes between bestlen-7 and bestlen, which is 8 bytes or 64 bits.
bestlen-3 actually means bytes between bestlen-3 and bestlen, which is 4 bytes or 32 bits.
bestlen-1 actually means bytes between bestlen-1 and bestlen, which is 2 bytes or 16 bits.

Reply:
Ah, that makes sense now :)

Reply:
It took me a while to decipher, because normally programmers use 0-based index, but length starts with 1 when we are dealing something that isn't NULL-terminated as when start and end are the same memory location, it counts as length of 1.

Reply:
Yeah, we are positioning the read so it reads past 7 bytes of best_len and 1 extra byte for next best length.